File: domains/org/src/Entities/Role/RoleType.cs
using Base.Entities;

namespace Org.Entities;

public class RoleType : ParameterBase
{
  public string Name { get; set; } = null!;

  public List<Role> Roles { get; set; } = new();

  public RoleType? Parent { get; set; }
  public List<RoleType> Children { get; set; } = new();

  public List<CapabilityType> CapabilityTypes { get; set; } = new();
}

File: webapp/src/Client/RoleType.ts
import { CapabilityType } from "./CapabilityTypes";
import { useCRUD, UseCRUDOptionalProps } from "./CRUD/CRUD";

export interface RoleType {
  id: number;
  name: string;
  active_from_date: string;
  retired_at_date?: string;
  parent_id?: number;
  parent?: RoleType;
  children: RoleType[];
  capability_types: CapabilityType[];
}

export interface MutateRoleType {
  name: string;
  retired_at_date?: string;
  parent_id?: number;
}

export interface UseRoleTypeProps extends UseCRUDOptionalProps { }

export function useRoleType(props?: UseRoleTypeProps) {
  const {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  } = useCRUD<RoleType, MutateRoleType>({
    path: "/api/org/roletypes",
    ...props,
  });

  return {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  };
}
